packages/query/src/block-processor.ts

```diff
@@ -0,0 +1,390 @@
+  private readonly fullViewingKey: string;
+  private readonly querier: RootQuerier;
+  private readonly indexedDb: IndexedDbInterface;
+  private readonly viewServer: ViewServerInterface;
+  private readonly abortController: AbortController = new AbortController();
+  private blockSyncPromise: Promise<void> | undefined;
+
+  constructor({ indexedDb, viewServer, querier, fullViewingKey }: QueryClientProps) {
+    this.fullViewingKey = fullViewingKey;
+    this.indexedDb = indexedDb;
+    this.viewServer = viewServer;
+    this.querier = querier;
+  }
+
+  // If syncBlocks() is called multiple times concurrently,
+  // they'll all wait for the same promise rather than each starting their own sync process.
+  async syncBlocks(): Promise<void> {
+    if (!this.blockSyncPromise) {
+      this.blockSyncPromise = this.autoRetrySync();
+    }
+    return this.blockSyncPromise;
+  }
+
+  // After a failure, retrying the sync is critical. An exponential-backoff is used.
+  async autoRetrySync() {
+    try {
+      await backOff(() => this.syncAndStore(), {
+        maxDelay: 30_000, // 30 seconds
+        retry: async error => {
+          await this.viewServer.resetTreeToStored();
+          if (isAbortSignal(error)) return false;
+          console.error('Syncing error', error);
+          return true;
+        },
+      });
+    } catch {
+      // Was signaled to abort, can swallow this error
+    }
+  }
+
+  stopSync() {
+    this.abortController.abort();
+  }
+
+  private async syncAndStore() {
+    const lastBlockSynced = await this.indexedDb.getLastBlockSynced();
+    const startHeight = lastBlockSynced ? lastBlockSynced + 1n : 0n;
+    const latestBlockHeight = await this.querier.tendermint.latestBlockHeight();
+
+    // this is an indefinite stream of the (compact) chain from the network
+    // intended to run continuously
+    for await (const compactBlock of this.querier.compactBlock.compactBlockRange({
+      startHeight,
+      keepAlive: true,
+      abortSignal: this.abortController.signal,
+    })) {
+      if (compactBlock.fmdParameters)
+        await this.indexedDb.saveFmdParams(compactBlock.fmdParameters);
+
+      if (compactBlock.gasPrices) await this.indexedDb.saveGasPrices(compactBlock.gasPrices);
+
+      // wasm view server scan
+      // - decrypts new notes
+      // - decrypts new swaps
+      // - updates idb with advice
+      //const shouldStoreProgress = await this.viewServer.scanBlock(compactBlock);
+      const wasmWantsFlush = await this.viewServer.scanBlock(compactBlock);
+
+      // flushing is slow, avoid it until
+      // - wasm says
+      // - every 1000th block
+      // - every block at tip
+      const flushReasons = {
+        wasmWantsFlush,
+        interval: compactBlock.height % 1000n === 0n,
+        new: compactBlock.height > latestBlockHeight,
+      };
+
+      let compactScan;
+      if (Object.values(flushReasons).some(Boolean)) {
+        compactScan = this.viewServer.flushUpdates();
```

Comment:
I've ignored refactoring before, but why `compactScan`?   
flushUpdates() returns updates, and updates are stored in this variable

Reply:
renamed